core/indexer/dataDispatcher.go

```diff
@@ -54,17 +58,32 @@ func (d *dataDispatcher) startWorker(ctx context.Context) {
 			log.Debug("dispatcher's go routine is stopping...")
 			return
 		case wi := <-d.chanWorkItems:
-			d.doWork(wi)
+			d.doWork(wi, 0)
 		}
 	}
 }
 
 // Close will close the endless running go routine
 func (d *dataDispatcher) Close() error {
+	start := time.Now()
 	if d.cancelFunc != nil {
 		d.cancelFunc()
 	}
 
+	d.wasClosed.Set()
+	close(d.chanWorkItems)
+	for wi := range d.chanWorkItems {
+		if time.Since(start) > closeTimeout {
+			log.Warn("cannot write all items from the queue",
+				"error", "timeout",
+			)
+			return nil
+		}
+
+		remainingTime := closeTimeout - time.Since(start)
```

Comment:
what happens if remainingTime will drop to 0 because the remainder jobs did not end? It will make the doWork function wait indefinitely. Do we have protection for negative values?

Reply:
Added extra check for the negative values. Also the case when the remaining time can be 0 was treated.